.github/workflows/release-cycle.yml

```diff
@@ -0,0 +1,178 @@
+
+jobs:
+  state:
+    name: Check state
+    permissions:
+      pull-requests: read
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up environment
+        uses: ./.github/actions/setup
+      - id: state
+        name: Get state
+        uses: actions/github-script@v6
+        with:
+          result-encoding: string
+          script: require('./scripts/release/workflow/state.js')({ github, context, core })
+    outputs:
+      # Job Flags
+      start: ${{ steps.state.outputs.start }}
+      changesets: ${{ steps.state.outputs.changesets }}
+      promote: ${{ steps.state.outputs.promote }}
+      publish: ${{ steps.state.outputs.publish }}
+      merge: ${{ steps.state.outputs.merge }}
+
+      # Global variables
+      is_prerelease: ${{ steps.state.outputs.is_prerelease }}
+      version: ${{ steps.state.outputs.version }}
+
+  start:
+    needs: state
+    name: Start new release branch with release candidate
+    permissions:
+      contents: write
+    if: needs.state.outputs.start == 'true'
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up environment
+        uses: ./.github/actions/setup
+      - run: bash scripts/git-user-config.sh
+      - name: Create branch with release candidate
+        run: bash scripts/release/workflow/start.sh
+
+  promote:
+    needs: state
+    name: Promote to final release
+    permissions:
+      contents: write
+    if: needs.state.outputs.promote == 'true'
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up environment
+        uses: ./.github/actions/setup
+      - run: bash scripts/git-user-config.sh
+      - name: Exit prerelease state if needed
+        run: bash scripts/release/workflow/exit-prerelease.sh
+        env:
+          PRERELEASE: ${{ needs.state.outputs.is_prerelease }}
+
+  changesets:
+    needs: [state, start, promote]
+    name: Changesets PR
+    permissions:
+      contents: write
+      pull-requests: write
+    if: ${{ always() && needs.state.outputs.changesets == 'true' }}
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0 # To get all tags
+      - name: Set up environment
+        uses: ./.github/actions/setup
+      - name: Set release title
+        uses: actions/github-script@v6
+        with:
+          result-encoding: string
+          script: require('./scripts/release/workflow/set-changesets-pr-title.js')({ core })
+      - name: Create PR
+        uses: changesets/action@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          version: npm run version
+          title: ${{ env.TITLE }}
+          commit: ${{ env.TITLE }}
+          body: | # Wait for support on this
+            This is an automated PR for releasing ${{ github.repository }}
+            Check [CHANGELOG.md](${{ github.repository }}/CHANGELOG.md)
+
+  pack:
```

Comment:
I don't think this needs to be a separate job! It's okay if `npm pack` executes in the "npm _deployment_ environment". My comment was that the command should not have `NPM_TOKEN` in the "execution" environment. Sorry about the confusion, I realize now that environment has these two meanings.

It could be something like:

```yaml
- id: pack
  run: bash scripts/release/workflow/pack.sh
- run: npm publish "$PACKAGE" --dist-tag="$TAG"
  env:
    NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
    PACKAGE: ${{ jobs.pack.outputs.tarball }}
    TAG: ${{ jobs.pack.outputs.tag }}
```

This is ideal because we see exactly where `NPM_TOKEN` is going.

Reply:
Fixed in [this commit](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3915/commits/c065664be4e22df0b240e21253b29388b78319b6)